x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java

```diff
@@ -1060,9 +1076,171 @@ private Expression propagate(And and) {
+                        Integer comp = BinaryComparison.compare(eq.right().fold(), neq.right().fold());
+                        if (comp != null) {
+                            if (comp == 0) {
+                                return FALSE; // clashing and conflicting: a = 1 AND a != 1
+                            } else {
+                                iter.remove(); // clashing and redundant: a = 1 AND a != 2
+                                changed = true;
+                            }
+                        }
+                    }
+                }
+
+                // evaluate all inequalities against the Equal
+                for (Iterator<BinaryComparison> iter = frInequalities.iterator(); iter.hasNext(); ) {
+                    BinaryComparison bc = iter.next();
+                    if (! eq.left().semanticEquals(bc.left())) {
+                        continue;
+                    }
+
+                    Integer compare = BinaryComparison.compare(eqValue, bc.right().fold());
+                    if (compare == null) {
+                        continue;
+                    }
+
+                    if (bc instanceof LessThan || bc instanceof LessThanOrEqual) { // a = 2 AND a </<= ?
+                        if ((compare == 0 && bc instanceof LessThan) || // a = 2 AND a < 2
+                            0 < compare) { // a = 2 AND a </<= 1
+                            return FALSE;
+                        }
+                    } else if (bc instanceof GreaterThan || bc instanceof GreaterThanOrEqual) { // a = 2 AND a >/>= ?
+                        if ((compare == 0 && bc instanceof GreaterThan) || // a = 2 AND a > 2
+                            compare < 0) { // a = 2 AND a >/>= 3
+                            return FALSE;
+                        }
+                    }
+
+                    iter.remove();
+                    changed = true;
+                }
+            }
+
+            return changed ? Predicates.combineAnd(CollectionUtils.combine(exps, frEquals, frNotEquals, frInequalities, ranges)) : and;
+        }
+
+        // combine disjunction:
+        // a = 2 OR a > 3 -> nop; a = 2 OR a > 1 -> a > 1
+        // a = 2 OR a < 3 -> a < 3; a = 2 OR a < 1 -> nop
+        // a = 2 OR 3 < a < 5 -> nop; a = 2 OR 1 < a < 3 -> 1 < a < 3; a = 2 OR 0 < a < 1 -> nop
+        // a = 2 OR a != 5 -> TRUE; a = 2 OR a = 5 -> nop
+        private Expression propagate(Or or) {
+            List<Expression> exps = new ArrayList<>();
+            List<Equals> frEquals = new ArrayList<>(); // foldable right term Equals
+            List<NotEquals> frNotEquals = new ArrayList<>(); // foldable right term NotEquals
+            List<Range> ranges = new ArrayList<>();
+            List<BinaryComparison> frInequalities = new ArrayList<>(); // foldable right term (=limit) BinaryComparision
+
+            // split expressions by type
+            for (Expression ex : Predicates.splitOr(or)) {
```

Comment:
Isn't this handled already by the folding rule  + BooleanSimplification?

Reply:
I believe it's not, no.
(But I'll happily be corrected.)

Reply:
Or at:
https://github.com/elastic/elasticsearch/blob/54c1a59650833361edb85d3a7e3bdb933d9571ba/x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java#L838
And at: 
https://github.com/elastic/elasticsearch/blob/54c1a59650833361edb85d3a7e3bdb933d9571ba/x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java#L809

That is each individual and/or is evaluated if possible

Reply:
I had a look at that, but the simplification added by the code isn't covered there. Maybe that's what the [`// TODO: eliminate conjunction/disjunction`](https://github.com/elastic/elasticsearch/blob/54c1a59650833361edb85d3a7e3bdb933d9571ba/x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java#L876) was hinting to?
 
(BTW and related to the last suggestions below, the comments [`(a || b) && (a || c) => a && (b || c)`](https://github.com/elastic/elasticsearch/blob/54c1a59650833361edb85d3a7e3bdb933d9571ba/x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java#L817) and [`(a && b) || (a && c) => a || (b & c)`](https://github.com/elastic/elasticsearch/blob/54c1a59650833361edb85d3a7e3bdb933d9571ba/x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java#L855) will also need fixing (the code does the right thing, though) -- I'll address this too.